---
title: "Fig3: H-MAGMA analysis"
output: html_notebook
---

# Setup
```{r include=FALSE}
# library(org.Hs.eg.db)
require(ACTIONet)
require(stringr)
require(ComplexHeatmap)


results.path = "~/results"
input.path = "~/results/input"
dataset.path = "~/results/datasets"
tables.path = "~/results/tables"
figures.path = "~/results/figures"


```

## Enrichment function
```{r}
assess.genesets <-function (arch.gs, terms.gs, N, min.pval = 1e-100, correct = "none"){
    shared = t(sapply(terms.gs, function(gs1) {
        sapply(arch.gs, function(gs2) {
            nn = intersect(gs1, gs2)
        })
    }))
    colnames(shared) = names(arch.gs)
    GS.sizes = sapply(terms.gs, length)
    logPvals.out = sapply(1:ncol(shared), function(i) {
        gg = shared[, i]
        x = as.numeric(sapply(gg, length))
        n.sample = length(arch.gs[[i]])
        n.success = as.numeric(GS.sizes)
        v = rep(1, length(x))
        min.overlap = n.success * n.sample/N
        idx = which(x >= min.overlap)
        if (length(idx) == 0) 
            return(v)
        v[idx] = (phyper(x[idx]-1, n.sample, N-n.sample, n.success[idx], lower.tail = F))
        # v[idx] = HGT_tail(population.size = N, success.count = n.success[idx], 
        #     sample.size = n.sample, observed.success = x[idx])
        return(v)
    })
    if(correct == "global") {
      logPvals.out = matrix(p.adjust(logPvals.out, "fdr"), nrow = nrow(logPvals.out))
    } else if(correct == "local") {
      logPvals.out = apply(logPvals.out, 2, function(x) p.adjust(x, "fdr"))
    }
    rownames(logPvals.out) = names(terms.gs)
    colnames(logPvals.out) = names(arch.gs)
    return(-log10(Matrix::t(logPvals.out)))
}
```

# Load primary datasets
```{r, eval = T}
ACTIONet_summary = readr::read_rds(file.path(dataset.path, "ACTIONet_summary_filtered_individuals.rds"))
pb.logcounts = readr::read_rds(file.path(dataset.path, "PB_mean_logcounts_final.RDS"))

color.df = readRDS(file.path(dataset.path, "celltype_colors.rds"))
colors = color.df$color
names(colors) = color.df$celltype

```


# Load DE results
```{r, eval = T}
resDE = readr::read_rds(file.path(dataset.path, "Cohort_specific_DE_results.rds"))
filtered.tables = readr::read_rds(file.path(dataset.path, "Cohort_specific_DE_results_filtered.rds"))
combined.analysis.tables = readr::read_rds(file.path(dataset.path, "meta_analysis_results.rds"))

DE.new = readRDS(file.path(dataset.path, "DE_genes_pseudobulk.rds"))
Up.genes = DE.new$Up.genes
Down.genes = DE.new$Down.genes
DE.sc = DE.new$DE.sc
# ordered.celltypes = rownames(X)[order(apply(X, 1, sum), decreasing = T)]

```

## Load significant variants and mapped genes
```{r}
PGC3.finemapped.genes.tbl = read.table(file.path(input.path, "Prioritised_PGC3_SZ_Genes.csv"), sep = "\t", header = T)
PGC3.loci = read.table(file.path(input.path, "PGC3_SZ_significant_loci.csv"), sep = "\t", header = T)

associated.genes = PGC3.loci$ENSEMBL.genes..all..clear.names.


PGC3.all.genes.raw = sort(unique(unlist(sapply(PGC3.loci$ENSEMBL.genes..all..clear.names., function(str) {
  if(str == "-") {
    return("-")
  }
  gs = str_split(str, ",")[[1]]
  
  return(gs)
}))))

PGC3.all.genes = intersect(PGC3.all.genes.raw, rownames(DE.sc))


```


# Compute transcriptional deviation scores
```{r}
common.genes = intersect(rownames(DE.new$DE.sc), rownames(pb.logcounts))
X.DE = DE.new$DE.sc[common.genes, ]
Z.DE = scale(X.DE)

idx = match(colnames(Z.DE), names(assays(pb.logcounts)))
scores = sapply(1:ncol(Z.DE), function(i) {
  print(i)
  
	Z = scale(assays(pb.logcounts)[[idx[[i]]]][common.genes, ])
	Z[is.na(Z)] = 0

	Z.ortho = orthoProject(Z, Matrix::rowMeans(Z))
	Z.ortho[is.na(Z.ortho)] = 0

	cc = (Matrix::t(Z.ortho) %*% Z.DE[, i]) / nrow(Z.DE)
})
colnames(scores) = colnames(Z.DE)
rownames(scores) = colnames(pb.logcounts)

# TPS = fast_row_sums(scores[, grepl("Ex-|In-", colnames(scores)) & !grepl("NRGN", colnames(scores))])
TPS = Matrix::rowSums(scores)
# TPS = fast_row_sums(scores[, grepl("Ex-", colnames(scores)) & !grepl("NRGN", colnames(scores))])
cor(TPS, pb.logcounts$A17.signature, method = "spearman")

pb.logcounts$TPS = TPS

```
```{r}
metadata.tbl = as.data.frame(colData(pb.logcounts))[, -c(29:31, 42:44)]

write.table(metadata.tbl, file.path(tables.path, "sample_metadata.tsv"), sep = "\t", row.names = F, col.names = T, quote = F)

readr::write_rds(metadata.tbl, file.path(dataset.path, "sample_metadata.rds"))


```




```{r}
# form = ~ Phenotype + Batch + PMI + Gender + Age + Benzodiazepines + Anticonvulsants + AntipsychTyp + AntipsychAtyp + Antidepress + SampleQuality
# 
# resDE = lapply( levels(pb.logcounts.filtered$Cohort), function(chrt){
# 
# 	keep.ids = colnames(pb.logcounts.filtered)[pb.logcounts.filtered$Cohort == chrt]
# 
# 	pb.logcounts.filtered_sub = pb.logcounts.filtered[,keep.ids]
#   sample.metadata = droplevels(data.frame(colData(pb.logcounts.filtered_sub)))
# 	design.mat <- model.matrix(form, data = sample.metadata)
# 	colnames(design.mat)[1] = c("Intercept")
# 
# 	contrast.mat <- makeContrasts(contrasts = "PhenotypeSZ", levels = design.mat)
# 
# 	df = pbDS(pb.logcounts.filtered_sub, method = "limma-trend", min_cells = 5, design = design.mat, contrast =  contrast.mat, filter = "both")
# 	
# })
# names(resDE) = levels(colData(pb.logcounts.filtered)$Cohort)
# 
# readr::write_rds(resDE, file.path(dataset.path, "Cohort_specific_DE_results.rds"))
# 


colnames(metadata.tbl)

metadata.tbl = colData(pb.logcounts)
out <-lm(A17.signature~TPS+Phenotype+umis+mito_perc+Age+Gender+PMI+ Benzodiazepines + Anticonvulsants + AntipsychTyp + AntipsychAtyp + Antidepress, data = metadata.tbl)
summary(out)



# metadata.tbl = colData(pb.logcounts)
# out <-lm(A17.signature~TPS+Phenotype+umis+mito_perc+Age+Gender+PMI+ Benzodiazepines + Anticonvulsants + AntipsychTyp + AntipsychAtyp + Antidepress + Anticonvulsants*Phenotype, data = metadata.tbl)
# summary(out)


sort(p.adjust(summary(out)$coefficients[-1,"Pr(>|t|)"], "fdr"))


```

# Comparison with Phenotype
## PRS
```{r}
sample.df = data.frame(Label = pb.logcounts$ID, Phenotype = pb.logcounts$Phenotype, TPS = scale(pb.logcounts$TPS), A17 = scale(pb.logcounts$A17.signature), PRS = scale(pb.logcounts$PRS))
sample.df = sample.df[pb.logcounts$POP.EL3SD == "EUR", ]


require(ggstatsplot)
gg = ggstatsplot::ggbetweenstats(
    data = sample.df,
    x = Phenotype,
    y = PRS,
    xlab = "Phenotype",
    ylab = "Polygenic Risk Score (PRS)",
    pairwise.display = "significant", # display only significant pairwise comparisons
    p.adjust.method = "fdr", # adjust p-values for multiple tests using this method
    ggtheme = ggthemes::theme_tufte(),
    outlier.tagging = TRUE,
    outlier.label = Label
  ) + scale_color_manual(values = c("darkgrey", "red"))
  
pdf(file.path(figures.path, "PRS_vs_phenotype_EUR.pdf"))
plot(gg)
dev.off()




sample.df$LabelPRS = ""
mask = ((sample.df$TPS < -1) & (sample.df$Phenotype == "SZ")) | ((sample.df$TPS >1) & (sample.df$Phenotype != "SZ"))
sample.df$LabelPRS[mask] = as.character(sample.df$Label[mask])

require(ggpubr)
gg = ggscatter(sample.df, x = "TPS", y = "PRS", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "LabelPRS", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson")#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "TPS_vs_PRS.pdf"), width = 6, height = 6)
print(gg)
dev.off()


```


# Comparison with Phenotype
## TPS
```{r}
sample.df = data.frame(Label = pb.logcounts$ID, Phenotype = pb.logcounts$Phenotype, TPS = scale(pb.logcounts$TPS), A17 = scale(pb.logcounts$A17.signature), A11 = scale(pb.logcounts$A11.signature), A29 = scale(pb.logcounts$A29.signature))

require(ggstatsplot)
gg = ggstatsplot::ggbetweenstats(
    data = sample.df,
    x = Phenotype,
    y = TPS,
    xlab = "Phenotype",
    ylab = "Transcriptional Pathology Score (TPS)",
    pairwise.display = "significant", # display only significant pairwise comparisons
    p.adjust.method = "fdr", # adjust p-values for multiple tests using this method
    ggtheme = ggthemes::theme_tufte(),
    outlier.tagging = TRUE,
    outlier.label = Label
  ) + scale_color_manual(values = c("darkgrey", "red"))
  
pdf(file.path(figures.path, "TPS_vs_phenotype.pdf"))
plot(gg)
dev.off()



require(ggstatsplot)
gg = ggstatsplot::ggbetweenstats(
    data = sample.df,
    x = Phenotype,
    y = A29,
    xlab = "Phenotype",
    ylab = "A29",
    pairwise.display = "significant", # display only significant pairwise comparisons
    p.adjust.method = "fdr", # adjust p-values for multiple tests using this method
    ggtheme = ggthemes::theme_tufte(),
    outlier.tagging = TRUE,
    outlier.label = Label
  ) + scale_color_manual(values = c("darkgrey", "red"))
  
pdf(file.path(figures.path, "A29_vs_phenotype.pdf"))
plot(gg)
dev.off()

require(ggstatsplot)
gg = ggstatsplot::ggbetweenstats(
    data = sample.df,
    x = Phenotype,
    y = A17,
    xlab = "Phenotype",
    ylab = "A17",
    pairwise.display = "significant", # display only significant pairwise comparisons
    p.adjust.method = "fdr", # adjust p-values for multiple tests using this method
    ggtheme = ggthemes::theme_tufte(),
    outlier.tagging = TRUE,
    outlier.label = Label
  ) + scale_color_manual(values = c("darkgrey", "red"))
  
pdf(file.path(figures.path, "A17_vs_phenotype.pdf"))
plot(gg)
dev.off()

require(ggstatsplot)
gg = ggstatsplot::ggbetweenstats(
    data = sample.df,
    x = Phenotype,
    y = A11,
    xlab = "Phenotype",
    ylab = "A11",
    pairwise.display = "significant", # display only significant pairwise comparisons
    p.adjust.method = "fdr", # adjust p-values for multiple tests using this method
    ggtheme = ggthemes::theme_tufte(),
    outlier.tagging = TRUE,
    outlier.label = Label
  ) + scale_color_manual(values = c("darkgrey", "red"))
  
pdf(file.path(figures.path, "A11_vs_phenotype.pdf"))
plot(gg)
dev.off()



```

# Compute enrichmen of SZ phenotype among top-ranked TPS scores
```{r}
require(fgsea)

SZ.inds = list(SZ = as.character(pb.logcounts$ID[which(pb.logcounts$Phenotype == "SZ")]))

v = pb.logcounts$TPS
names(v) = pb.logcounts$ID

all.enrichment = fgsea::fgsea(SZ.inds, v)
pdf(file.path(figures.path, "Supp", "TPS_all_vs_Pheno_GSEA.pdf"), height = 4)
plotEnrichment(SZ.inds$SZ, v)+ labs(title=sprintf("TPS-vs-Phenotype (adjusted p-val = %.2e)", all.enrichment$padj))
dev.off()


v = pb.logcounts$A11.signature
names(v) = pb.logcounts$ID

all.enrichment = fgsea::fgsea(SZ.inds, v, scoreType = "pos")
pdf(file.path(figures.path, "Supp", "A11_all_vs_Pheno_GSEA.pdf"), height = 4)
plotEnrichment(SZ.inds$SZ, v)+ labs(title=sprintf("A11-vs-Phenotype (adjusted p-val = %.2e)", all.enrichment$padj))
dev.off()



v = pb.logcounts$A17.signature
names(v) = pb.logcounts$ID

all.enrichment = fgsea::fgsea(SZ.inds, v, scoreType = "pos")
pdf(file.path(figures.path, "Supp", "A17_all_vs_Pheno_GSEA.pdf"), height = 4)
plotEnrichment(SZ.inds$SZ, v)+ labs(title=sprintf("A17-vs-Phenotype (adjusted p-val = %.2e)", all.enrichment$padj))
dev.off()



v = pb.logcounts$A29.signature
names(v) = pb.logcounts$ID

all.enrichment = fgsea::fgsea(SZ.inds, v, scoreType = "pos")
pdf(file.path(figures.path, "Supp", "A29_all_vs_Pheno_GSEA.pdf"), height = 4)
plotEnrichment(SZ.inds$SZ, v)+ labs(title=sprintf("A29-vs-Phenotype (adjusted p-val = %.2e)", all.enrichment$padj))
dev.off()




```

```{r}
arch.pheno.annot = annotate.archetypes.using.labels(t(ACTIONet_summary$H_unified), ACTIONet_summary$metadata$Phenotype)


ComplexHeatmap::Heatmap(arch.pheno.annot$Enrichment)

cor.test(pb.logcounts$A29.signature, pb.logcounts$umis)
cor.test(pb.logcounts$SZTR.mean, pb.logcounts$umis)



```

```{r}
ComplexHeatmap::Heatmap(arch.pheno.annot$Enrichment)

```


```{r}
x = pb.logcounts$A17.signature
z = (x - median(x)) / mad(x)

mask = pb.logcounts$Cohort == "McLean"
cor.test(pb.logcounts$A17.signature, pb.logcounts$TPS)
cor.test(pb.logcounts$A17.signature[mask], pb.logcounts$TPS[mask])
cor.test(pb.logcounts$A17.signature[!mask], pb.logcounts$TPS[!mask])


# 	Pearson's product-moment correlation
# 
# data:  pb.logcounts$A17.signature and pb.logcounts$TPS
# t = -10.511, df = 138, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.7497761 -0.5632232
# sample estimates:
#        cor 
# -0.6668185 
# 
# 
# 	Pearson's product-moment correlation
# 
# data:  pb.logcounts$A17.signature[mask] and pb.logcounts$TPS[mask]
# t = -6.7751, df = 46, p-value = 1.983e-08
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.8251658 -0.5287768
# sample estimates:
#        cor 
# -0.7067297 
# 
# 
# 	Pearson's product-moment correlation
# 
# data:  pb.logcounts$A17.signature[!mask] and pb.logcounts$TPS[!mask]
# t = -7.277, df = 90, p-value = 1.226e-10
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.7232834 -0.4613185
# sample estimates:
#        cor 
# -0.6086289 



```



# Filter samples
```{r}
ncells = sapply(int_colData(pb.logcounts)$n_cells, as.numeric)
rownames(ncells) = names(assays(pb.logcounts))

cs = Matrix::colSums(ncells)
ncells.freq = 100*scale(ncells, center = F, scale = cs)
Ex.perc = (Matrix::colSums(ncells.freq[grepl("^Ex", rownames(ncells.freq)) & !grepl("^Ex-NRGN", rownames(ncells.freq)), ]))

mask = (Ex.perc >= 10) & (Ex.perc <= 80)


pb.logcounts.filtered = pb.logcounts [, mask]

```


```{r}
require(ggpubr)
df = data.frame(TPS = scale(pb.logcounts.filtered$TPS.Neuro), SZTR = scale(pb.logcounts.filtered$A17.signature), Phenotype = pb.logcounts.filtered$Phenotype, ID = pb.logcounts.filtered$ID)
df$Label = ""
TPS.mask = ((df$TPS >1) & (df$Phenotype == "CON")) | ((df$TPS < -1) & (df$Phenotype == "SZ"))
df$Label[TPS.mask] = as.character(df$ID[TPS.mask])


gg = ggscatter(df, x = "TPS", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "TPS_vs_SZTR.pdf"), width = 6, height = 6)
print(gg)
dev.off()


```


```{r}
require(ggpubr)
df = data.frame(TPS = scale(pb.logcounts.filtered$TPS.Neuro), SZTR = scale(pb.logcounts.filtered$A17.signature), Phenotype = pb.logcounts.filtered$Phenotype, ID = pb.logcounts.filtered$ID, Cohort = pb.logcounts.filtered$Cohort)
df$Label = ""
TPS.mask = ((df$TPS >1) & (df$Phenotype == "CON")) | ((df$TPS < -1) & (df$Phenotype == "SZ"))
df$Label[TPS.mask] = as.character(df$ID[TPS.mask])

df1 = df[df$Cohort == "McLean", ]

gg = ggscatter(df1, x = "TPS", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "TPS_vs_SZTR_McLean.pdf"), width = 6, height = 6)
print(gg)
dev.off()


df2 = df[df$Cohort != "McLean", ]

gg = ggscatter(df2, x = "TPS", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "TPS_vs_SZTR_MtSinai.pdf"), width = 6, height = 6)
print(gg)
dev.off()


```
```{r}
require(ggpubr)
df = data.frame(SZTR = scale(pb.logcounts.filtered$A17.signature), mito = scale(pb.logcounts.filtered$mito_perc), genes = scale(pb.logcounts.filtered$genes), umis = scale(pb.logcounts.filtered$umis), Phenotype = pb.logcounts.filtered$Phenotype)

df$Label = ""
TPS.mask = ((df$TPS >1) & (df$Phenotype == "CON")) | ((df$TPS < -1) & (df$Phenotype == "SZ"))
df$Label[TPS.mask] = as.character(df$ID[TPS.mask])

gg = ggscatter(df, x = "mito", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "Mito_vs_SZTR.pdf"), width = 6, height = 6)
print(gg)
dev.off()


gg = ggscatter(df, x = "genes", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "Genes_vs_SZTR.pdf"), width = 6, height = 6)
print(gg)
dev.off()

gg = ggscatter(df, x = "umis", y = "SZTR", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "UMI_vs_SZTR.pdf"), width = 6, height = 6)
print(gg)
dev.off()



```



```{r}
require(ggpubr)
df = data.frame(A29 = scale(pb.logcounts.filtered$A29.signature), mito = scale(pb.logcounts.filtered$mito_perc), genes = scale(pb.logcounts.filtered$genes), umis = scale(pb.logcounts.filtered$umis), Phenotype = pb.logcounts.filtered$Phenotype)

df$Label = ""
TPS.mask = ((df$TPS >1) & (df$Phenotype == "CON")) | ((df$TPS < -1) & (df$Phenotype == "SZ"))
df$Label[TPS.mask] = as.character(df$ID[TPS.mask])

gg = ggscatter(df, x = "mito", y = "A29", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "Mito_vs_A29.pdf"), width = 6, height = 6)
print(gg)
dev.off()


gg = ggscatter(df, x = "genes", y = "A29", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "Genes_vs_A29.pdf"), width = 6, height = 6)
print(gg)
dev.off()

gg = ggscatter(df, x = "umis", y = "A29", 
          color = "Phenotype",
          palette = c("CON" = "lightgray", "SZ" = "red"),
          label = "Label", repel = TRUE,
          add = "reg.line",                                 # Add regression line
          conf.int = TRUE,                                  # Add confidence interval
          add.params = list(color = "blue",
                            fill = "lightgray")
          ) + geom_vline(xintercept = -1, linetype = "dashed", color="gray") + geom_vline(xintercept = 1, linetype = "dashed", color="gray")+
  stat_cor(method = "pearson", label.x = -0.8)#  +xlim(c(-2, 2)) + ylim(c(-2, 2))

pdf(file.path(figures.path, "Supp", "UMI_vs_A29.pdf"), width = 6, height = 6)
print(gg)
dev.off()

```

```{r}
data("gProfilerDB_human")
En = annotate.profile.using.markers(t(ACTIONet_summary$unified_feature_specificity), gProfilerDB_human$SYMBOL$`GO:BP`)

gProfileR::gprofiler()
```

```{r}
states = cbind(pb.logcounts.filtered$A7.signature, pb.logcounts.filtered$A11.signature, pb.logcounts.filtered$A17.signature, pb.logcounts.filtered$A29.signature)
stats = cbind(pb.logcounts.filtered$mito_perc, pb.logcounts.filtered$genes, pb.logcounts.filtered$umis, pb.logcounts.filtered$TPS)
colnames(states) = c("A7_Ex-NRGN", "A11_Ex-Mt", "A17_Ex-SZTR", "A29_In-Mt")
colnames(stats) = c("Mito", "Genes", "UMIs", "TPS")

pvals = t(apply(states, 2, function(x) apply(stats, 2, function(y) cor.test(x, y)$p.value)))
pvals.cor = matrix(p.adjust(pvals, method = "fdr"), nrow = ncol(states))
rownames(pvals.cor) = c("A7_Ex-NRGN", "A11_Ex-Mt", "A17_Ex-SZTR", "A29_In-Mt")
colnames(pvals.cor) = c("Mito", "Genes", "UMIs", "TPS")

CC = cor(states, stats)
ComplexHeatmap::Heatmap(sign(CC)*(-log10(pvals)))


# RbBu.pal = circlize::colorRamp2(seq(-1, 1, length.out = 7), pals::brewer.rdbu(7))


RbBu.pal = colorRampPalette(rev(RColorBrewer::brewer.pal(9, "RdBu")))(200)
# RbBu.pal = c(rep(RbBu.pal[1], length(RbBu.pal)*(sum(X < 1) / length(X))), RbBu.pal)


require(corrplot)
pdf(file.path(figures.path, "Cellstates_vs_stats.pdf"), width = 4, height = 4)
corrplot(CC, is.corr = T, method = "square", cl.length = 5, outline = T, sig.level = 0.01, p.mat = pvals.cor, insig = "label_sig", tl.col = "black", col = RbBu.pal)
dev.off()



```

```{r}
SynGo = readRDS("~/SynGO_release_20180731.rds")
SynGo.BP = SynGo$BP[sapply(SynGo$BP, length) >= 5]
SynGo.CC = SynGo$CC[sapply(SynGo$CC, length) >= 5]

v = ACTIONet_summary$unified_feature_specificity[, 17]
names(v) = rownames(ACTIONet_summary$unified_feature_specificity)
SZTR.SynGO.BP = fgsea::fgsea(SynGo.CC, v)
# postsynaptic endocytic zone (GO:0098843)
rownames(SZTR.SynGO.BP) = SZTR.SynGO.BP$pathway
SZTR.SynGO.BP$leadingEdge[SZTR.SynGO.BP$pathway == "postsynaptic endocytic zone (GO:0098843)"] # "PICK1"   "CALY"    "SH3GLB2" "CLTA"    "RNF216"  "SYNE1" -- padj = 0.0000000037

```




```{r}
v = ACTIONet_summary$unified_feature_specificity[, 17]
names(v) = rownames(ACTIONet_summary$unified_feature_specificity)

set.seed(0)
DE = c(DE.new$Up.genes, DE.new$Down.genes)
names(DE) = c(paste("Up", names(DE.new$Up.genes), sep = "_"), paste("Down", names(DE.new$Down.genes), sep = "_"))
DE.enrich = fgsea::fgsea(DE, v)
saveRDS(DE.enrich, "~/results/datasets/SZTR_vs_DE.rds")
# 
# DE.enrich = readRDS("~/results/datasets/SZTR_vs_DE.rds")
# DE.enrich2 = readRDS("~/results/datasets/SZTR_vs_DE2.rds")


mask = DE.enrich$padj < 0.05
cc = table(unlist(DE.enrich$leadingEdge[mask]))
shared.genes = sort(unique(names(cc)[cc > 1]))



# DE.enrich = DE.enrich[order(DE.enrich$pval), ]
# DE.enrich$pathway[DE.enrich$padj < 0.05]
# 
# DE.enrich2 = DE.enrich2[order(DE.enrich2$pval), ]
# DE.enrich2$pathway[DE.enrich2$padj < 0.05]


length(shared.genes)

shared.genes.GWAS = intersect(shared.genes, PGC3.all.genes) # "BCL11A"   "BRINP2"   "BSN"      "CES4A"    "CUX2"     "DGKI"     "GABBR2"   "GRAMD1B"  "GRIN2A"   "IGSF9B"   "KCNB1"    "MPPED2"   "NDST3"    "PPP1R13B"

v[shared.genes.GWAS] > -log10(0.05)

shared.genes.GWAS.gp = gProfileR::gprofiler(shared.genes.GWAS) #, hier_filtering = "moderate", src_filter = c("GO:BP"))

shared.genes.GWAS.gp$intersection[[1]] # "IGSF9B,CUX2,GABBR2,DGKI,KCNB1,BSN,GRIN2A"

require(ComplexHeatmap)
X = -DE.new$DE.sc[shared.genes.GWAS, ]
X[X < 0] = 0
mask = (abs(DE.new$logFC[shared.genes.GWAS, ]) > 0.1) & (DE.new$logPvals[shared.genes.GWAS, ] > -log10(0.05))
X[!mask] = 0

spec = as.numeric(scale(log1p(ACTIONet_summary$unified_feature_specificity[, 17])))
names(spec) = rownames(ACTIONet_summary$unified_feature_specificity)        
spec = spec[shared.genes.GWAS]
```

```{r}
require(org.Hs.eg.db)
HMAGMA.tbl = read.table("~/magma/hmagma/hmagmaAdultBrain__sz3/hmagmaAdultBrain__sz3.genes.out", header = T)
suppressWarnings(ids <- AnnotationDbi::mapIds(org.Hs.eg.db, keys = HMAGMA.tbl$GENE, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first"))
ids[is.na(ids)] = ""

HMAGMA.tbl$Symbol = ids
Hm = -log10(HMAGMA.tbl$P[match(shared.genes.GWAS, ids)])

perm = order(spec, decreasing = T)

pdf(file.path(figures.path, "SZTR_vs_GWAS_heatmap.pdf"), height = 4.5)
Heatmap(X[perm, ], cluster_rows = F, col = blues9, rect_gp = gpar(col = "black"), column_names_gp = gpar(col=colors), cluster_columns = F, name = "Differential", row_names_side = "left", show_row_dend = F) + Heatmap(as.matrix(spec[perm]), col = viridis::magma(11), rect_gp = gpar(col = "black"), name = "SZTR specificity", row_names_side = "left") + Heatmap(as.matrix(Hm[perm]), col = viridis::magma(11), rect_gp = gpar(col = "black"), name = "HMAGMA", row_names_side = "left")
dev.off()

```

```{r}
shared.genes.gp.filtered = gProfileR::gprofiler(shared.genes, hier_filtering = "moderate", src_filter = c("GO:BP", "GO:CC"))
readr::write_rds(shared.genes.gp.filtered, "~/results/datasets/SZTR_vs_DE_shared_genes_gProfiler_filtered.rds")



```

```{r}
shared.genes.gp = gProfileR::gprofiler(shared.genes) #, hier_filtering = "moderate", src_filter = c("GO:BP"))
readr::write_rds(shared.genes.gp, "~/results/datasets/SZTR_vs_DE_shared_genes_gProfiler.rds")

gsum = GOsummaries::gosummaries(list(SZTR = shared.genes)) #, hier_filtering = "moderate", src_filter = c("GO:BP"))
plot(gsum, fontsize = 8, filename = "~/results/figures/SZTR_vs_vs_DE_enrichment_goSum.pdf")

```

```{r}
require(org.Hs.eg.db)
SCHEMA = read.csv("~/results/input/SCHEMA.csv", sep = ",")

suppressWarnings(ids <- AnnotationDbi::mapIds(org.Hs.eg.db, keys = SCHEMA$Gene, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first"))
ids[is.na(ids)] = ""

SCHEMA$gene.name = ids
SCHEMA = SCHEMA[SCHEMA$gene.name != "", ]

v = -log10(SCHEMA$P.meta)
names(v) = SCHEMA$gene.name

DE = c(DE.new$Up.genes, DE.new$Down.genes)
names(DE) = c(paste("Up", names(DE.new$Up.genes), sep = "_"), paste("Down", names(DE.new$Down.genes), sep = "_"))
DE.SCHEMA.enrich = fgsea::fgsea(DE, v)

readr::write_rds(DE.SCHEMA.enrich, file.path(dataset.path, "SCHEMA_enrichment.rds"))



```


```{r}
v = -log10(SCHEMA$P.meta)
names(v) = SCHEMA$gene.name

DE.combined = lapply(1:length(Up.genes), function(i) sort(unique(union(Up.genes[[i]], Down.genes[[i]]))))
names(DE.combined) = names(Up.genes)

```


```{r}

z = scale(ACTIONet_summary$unified_feature_specificity)[, 17]
SZTR.genes = data.frame(Genes = shared.genes, SZTR_score = z[shared.genes])

selected.cts = sapply(str_split(DE.enrich2$pathway[DE.enrich2$padj < 0.05], "_"), function(l) paste(l[-1], collapse = "_"))
selected.cts = selected.cts[grep("^Ex|^In", selected.cts)]
DE.mask = as.data.frame(sapply(DE.new$Down.genes[selected.cts], function(gs) SZTR.genes$Genes %in% gs))
colnames(DE.mask) = selected.cts
DE.mask = DE.mask[, -7]

isGWAS = shared.genes %in% PGC3.all.genes
HMAGMA = -log10(HMAGMA.tbl$P[match(shared.genes, HMAGMA.tbl$Symbol)])
SCHEMA.scores = -log10(SCHEMA$P.meta[match(shared.genes, SCHEMA$gene.name)])
var.df = data.frame(isGWAS = isGWAS, HMAGMA = HMAGMA, SCHEMA = SCHEMA.scores)
  
  
SZTR.genes = cbind(SZTR.genes, DE.mask, var.df)

SZTR.genes = SZTR.genes[order(SZTR.genes$SZTR_score, decreasing = T), ]
write.table(SZTR.genes, file.path(tables.path, "SZTR_genes.tsv"), sep = "\t", row.names = F, col.names = T, quote = F)


```


## Export selected results
```{r}
# HMAGMA.Pvals = readr::read_rds(file.path(dataset.path, "HMAGMA_results_raw.rds"))

  selected.traits = c("hmagmaAdultBrain__sz3", "hmagmaAdultBrain__bip2", "hmagmaAdultBrain__mdd_without_23andMe", "hmagmaAdultBrain__adhd", "hmagmaAdultBrain__asd", "hmagmaAdultBrain__alz2noapoe")

trait.labels = c("Schizophrenia (SZ)", "Bipolar (BP)", "Depression (MDD)", "ADHD","Autism (ASD)", "Alzheimer (AD)")

X = HMAGMA.Pvals$UpAndDown[, selected.traits]
cts = intersect(color.df$celltype, rownames(X))
X = X[cts, ]
X = matrix(p.adjust(as.numeric(X), "fdr"), nrow = length(cts))
rownames(X) = cts
# X = apply(X, 2, function(p) p.adjust(p, "fdr"))
rownames(X) = intersect(names(colors), rownames(X))
X = -log10(X)
X[X < -log10(0.05)] = 0
colnames(X) = trait.labels
rownames(X) = cts


PurPal = colorRampPalette(RColorBrewer::brewer.pal(9, "Purples"))(200)
PurPal = c(rep(PurPal[1], length(PurPal)*(sum(X < 1) / length(X))), PurPal)

corrplot(X, is.corr = F, method = "pie", col = PurPal, cl.lim = c(0, 4), cl.length = 5, outline = T, sig.level = 0.05, p.mat = 10^(-X), insig = "blank", tl.col = "black")


```


```{r}
set.seed(0)
# names(DE) = c(paste("Up", names(DE.new$Up.genes), sep = "_"), paste("Down", names(DE.new$Down.genes), sep = "_"))
DE.combined.SCHEMA.enrich = fgsea::fgsea(DE.combined, v, eps = 0)

Y = as.matrix(-log10(DE.combined.SCHEMA.enrich$padj))
rownames(Y) = DE.combined.SCHEMA.enrich$pathway
colnames(Y) = "Schizophrenia (SCHEMA)"

Z = cbind(Y[intersect(names(colors), rownames(Y)), ], X[intersect(names(colors), rownames(X)), ])

require(corrplot)
pdf(file.path(figures.path, "SCHEMA_plus_HMAGMA.pdf"), width =7, height = 7)
corrplot(Z, is.corr = F, method = "pie", col = PurPal, cl.lim = c(0, 4), cl.length = 5, outline = T, sig.level = 0.05, p.mat = 10^(-Z), insig = "blank", tl.col = "black") 
dev.off()





readr::write_rds(DE.SCHEMA.enrich, file.path(dataset.path, "SCHEMA_enrichment.rds"))

```

```{r}
DE.enrich = readRDS("~/results/datasets/SZTR_vs_DE.rds")


mask = DE.enrich$padj < 0.05
cc = table(unlist(DE.enrich$leadingEdge[mask]))
shared.genes = sort(unique(names(cc)[cc > 1]))


SZTR.SCHEMA.enrich = fgsea::fgsea(list(SZTR = shared.genes), v)
readr::write_rds(SZTR.SCHEMA.enrich, file.path(dataset.path, "SCHEMA_enrichment_SZTE.rds"))

intersect(SZTR.SCHEMA.enrich$leadingEdge[[1]], PGC3.all.genes)


```

# Visualize
## Without PRS (All)
```{r}
ll.common = intersect(names(colors), colnames(scores))
selected.colors = colors[ll.common]

sub.scores = scores[mask, match(ll.common, colnames(scores))]

pheno = factor(pb.logcounts.filtered$Phenotype)
SZTR.mean = scale(pb.logcounts.filtered$A17.signature)
TPS = pb.logcounts.filtered$TPS
PRS = scale(pb.logcounts.filtered$PRS)
perm = order(TPS, decreasing = T)
names(TPS) = pb.logcounts.filtered$ID



pdf(file.path(figures.path, "TPS_heatmap_all.pdf"), height = 24, width = 6)
Heatmap(TPS[perm], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "TPS", rect_gp = gpar(col = "black")) + Heatmap(sub.scores[perm, grep("^Ex", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Ex", column_title =  "Ex", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^Ex", colnames(sub.scores))])) + Heatmap(sub.scores[perm, grep("^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "In", column_title = "In", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^In", colnames(sub.scores))])) + Heatmap(sub.scores[perm, !grepl("^Ex|^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Glial", column_title = "Glial", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[!grepl("^Ex|^In", colnames(sub.scores))])) + Heatmap(pheno[perm], col = c("lightgray", "red"), name = "Phenotype", rect_gp = gpar(col = "black")) + Heatmap(SZTR.mean[perm], name = "SZTR", rect_gp = gpar(col = "black"), col = inferno(500))
dev.off()




pdf(file.path(figures.path, "TPS_heatmap_all_neuro.pdf"), height = 24, width = 5)
Heatmap(TPS[perm], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "TPS", rect_gp = gpar(col = "black")) + Heatmap(sub.scores[perm, grep("^Ex", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Ex", column_title =  "Ex", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^Ex", colnames(sub.scores))])) + Heatmap(sub.scores[perm, grep("^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "In", column_title = "In", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^In", colnames(sub.scores))])) + Heatmap(pheno[perm], col = c("lightgray", "red"), name = "Phenotype", rect_gp = gpar(col = "black")) + Heatmap(SZTR.mean[perm], name = "SZTR", rect_gp = gpar(col = "black"), col = inferno(500))
dev.off()


```

### Only Eur + PRS
```{r}
ll.common = intersect(names(colors), colnames(scores))
selected.colors = colors[ll.common]

sub.scores = scores[mask, match(ll.common, colnames(scores))]


pheno = factor(pb.logcounts.filtered$Phenotype)
SZTR.mean = scale(pb.logcounts.filtered$A17.signature)
TPS = pb.logcounts.filtered$TPS.Neuro
PRS = scale(pb.logcounts.filtered$PRS)
perm = order(TPS, decreasing = T)
names(TPS) = pb.logcounts.filtered$ID

# require(viridis)
# pdf(file.path(figures.path, "TPS_heatmap_plus_PRS.pdf"), height = 24, width = 5)
# Heatmap(TPS[perm], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "TPS", rect_gp = gpar(col = "black")) + Heatmap(sub.scores[perm, grep("^Ex", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Ex", column_title =  "Ex", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^Ex", colnames(sub.scores))])) + Heatmap(sub.scores[perm, grep("^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "In", column_title = "In", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^In", colnames(sub.scores))])) + Heatmap(pheno[perm], col = c("lightgray", "red"), name = "Phenotype", rect_gp = gpar(col = "black")) + Heatmap(SZTR.mean[perm], name = "SZTR", rect_gp = gpar(col = "black"), col = magma(500)) + Heatmap(PRS[perm], name = "PRS", rect_gp = gpar(col = "black"))
# dev.off()






pdf(file.path(figures.path, "TPS_heatmap_plus_PRS_ext.pdf"), height = 24, width = 6)
Heatmap(TPS[perm], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "TPS", rect_gp = gpar(col = "black")) + Heatmap(sub.scores[perm, grep("^Ex", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Ex", column_title =  "Ex", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^Ex", colnames(sub.scores))])) + Heatmap(sub.scores[perm, grep("^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "In", column_title = "In", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^In", colnames(sub.scores))])) + Heatmap(sub.scores[perm, !grepl("^Ex|^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Glial", column_title = "Glial", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[!grepl("^Ex|^In", colnames(sub.scores))])) + Heatmap(pheno[perm], col = c("lightgray", "red"), name = "Phenotype", rect_gp = gpar(col = "black")) + Heatmap(SZTR.mean[perm], name = "SZTR", rect_gp = gpar(col = "black"), col = magma(500)) + Heatmap(PRS[perm], name = "PRS", rect_gp = gpar(col = "black"))
dev.off()


pdf(file.path(figures.path, "TPS_heatmap_plus_PRS_neuro_Ext.pdf"), height = 24, width = 5)
Heatmap(TPS[perm], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "TPS", rect_gp = gpar(col = "black")) + Heatmap(sub.scores[perm, grep("^Ex", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "Ex", column_title =  "Ex", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^Ex", colnames(sub.scores))])) + Heatmap(sub.scores[perm, grep("^In", colnames(sub.scores))], cluster_rows = F, cluster_columns = F, row_names_side = "left", name = "In", column_title = "In", rect_gp = gpar(col = "black"), column_names_gp = gpar(col = selected.colors[grep("^In", colnames(sub.scores))])) + Heatmap(pheno[perm], col = c("lightgray", "red"), name = "Phenotype", rect_gp = gpar(col = "black")) + Heatmap(SZTR.mean[perm], name = "SZTR", rect_gp = gpar(col = "black"), col = magma(500)) + Heatmap(PRS[perm], name = "PRS", rect_gp = gpar(col = "black"))
dev.off()




```


